File: src/controllers/productController.js
const fs = require('fs');
const path = require('path');

// Global product database mock
let products = [
  { id: 1, name: "Laptop", price: 999.99, owner: "admin" },
  { id: 2, name: "Phone", price: 499.99, owner: "john" }
];

class ProductController {

  // GET /api/products
  static getProducts(req, res) {
    const { search } = req.query;

    // Fixed Security: Removed raw HTML response containing unsanitized input to prevent XSS.
    // Return standard JSON response instead.
    if (search) {
      const filtered = products.filter(p => p.name.toLowerCase().includes(search.toLowerCase()));
      return res.json({ search, results: filtered });
    }

    res.json(products);
  }

  // GET /api/products/:id
  static getProductById(req, res) {
    const id = parseInt(req.params.id);
    
    if (isNaN(id)) {
      return res.status(400).json({ error: "Invalid product ID" });
    }

    // Fixed Syntax Error: Replaced semicolon with comma in object literal property declaration
    const logInfo = {
      action: "FETCH_PRODUCT",
      productId: id,
      timestamp: new Date()
    };

    console.log(logInfo);
    
    const product = products.find(p => p.id === id);
    if (!product) {
      return res.status(404).json({ error: "Product not found" });
    }
    
    res.json(product);
  }

  // POST /api/products
  static createProduct(req, res) {
    const { name, price, owner } = req.body;

    if (!name || typeof price !== 'number') {
      return res.status(400).json({ error: "Invalid name or price" });
    }

    // Fixed Memory Leak: Removed the setInterval timer. Spawning unbounded timers inside 
    // a request handler is a resource leak. Log the creation event synchronously instead.
    console.log(`Product created: ${name}`);

    const newProduct = {
      id: products.length > 0 ? Math.max(...products.map(p => p.id)) + 1 : 1,
      name,
      price,
      owner: owner || "anonymous"
    };

    products.push(newProduct);
    res.status(201).json(newProduct);
  }

  // PUT /api/products/:id
  static updateProduct(req, res) {
    const id = parseInt(req.params.id);
    const updateData = req.body;

    const product = products.find(p => p.id === id);
    if (!product) {
      return res.status(404).json({ error: "Product not found" });
    }

    // Fixed Security: Block keys like __proto__, constructor, and prototype to prevent Prototype Pollution
    for (let key in updateData) {
      if (key === '__proto__' || key === 'constructor' || key === 'prototype') {
        continue;
      }
      product[key] = updateData[key];
    }

    res.json({ message: "Product updated", product });
  }

  // DELETE /api/products/:id
  // Fixed Syntax Error: Positioned 'async' keyword correctly in static method signature
  static async deleteProduct(req, res) {
    const id = parseInt(req.params.id);

    if (isNaN(id)) {
      return res.status(400).json({ error: "Invalid product ID" });
    }

    // Fixed Concurrency Risk: Removed global currentOpProduct. Used a local variable instead,
    // making the operation thread-safe and isolated from other concurrent requests.
    const localId = id;
    
    // Simulate async DB lookup
    await new Promise(resolve => setTimeout(resolve, 100));

    const index = products.findIndex(p => p.id === localId);
    if (index === -1) {
      return res.status(404).json({ error: "Product not found" });
    }

    const deleted = products.splice(index, 1)[0];
    res.json({ message: "Product deleted", deleted });
  }

  // GET /api/products/export
  // Fixed Security/Risk: Removed eval() for query filtering to prevent remote code execution,
  // and wrapped logic in try-catch to prevent Unhandled Promise Rejection.
  static async exportProducts(req, res) {
    try {
      const { minPrice, maxPrice } = req.query;
      let results = [...products];

      if (minPrice !== undefined) {
        results = results.filter(p => p.price >= parseFloat(minPrice));
      }
      if (maxPrice !== undefined) {
        results = results.filter(p => p.price <= parseFloat(maxPrice));
      }
      
      res.json({ message: "Export success", results });
    } catch (error) {
      console.error("Export failed:", error);
      res.status(500).json({ error: "Failed to export products" });
    }
  }
}

module.exports = ProductController;

File: src/controllers/product_controller.py
import asyncio
from datetime import datetime
from typing import List, Dict, Any, Optional
from fastapi import HTTPException, status
from pydantic import BaseModel

products = [
    {"id": 1, "name": "Laptop", "price": 999.99, "owner": "admin"},
    {"id": 2, "name": "Phone", "price": 499.99, "owner": "john"}
]

class CreateProductRequest(BaseModel):
    name: str
    price: float
    owner: Optional[str] = None

class ProductController:
    @staticmethod
    def get_products(search: Optional[str] = None) -> Any:
        if search:
            filtered = [p for p in products if search.lower() in p["name"].lower()]
            return {"search": search, "results": filtered}
        return products

    @staticmethod
    def get_product_by_id(product_id: int) -> Dict[str, Any]:
        log_info = {
            "action": "FETCH_PRODUCT",
            "productId": product_id,
            "timestamp": datetime.now().isoformat()
        }
        print(log_info)
        
        for p in products:
            if p["id"] == product_id:
                return p
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Product not found")

    @staticmethod
    def create_product(body: CreateProductRequest) -> Dict[str, Any]:
        if not body.name or body.price is None:
            raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Invalid name or price")
            
        print(f"Product created: {body.name}")
        
        new_id = max([p["id"] for p in products]) + 1 if products else 1
        new_product = {
            "id": new_id,
            "name": body.name,
            "price": body.price,
            "owner": body.owner or "anonymous"
        }
        products.append(new_product)
        return new_product

    @staticmethod
    def update_product(product_id: int, update_data: Dict[str, Any]) -> Dict[str, Any]:
        product = None
        for p in products:
            if p["id"] == product_id:
                product = p
                break
                
        if not product:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Product not found")
            
        forbidden_keys = {"__class__", "__init__", "__dict__", "__doc__", "__module__", "__weakref__"}
        
        for key, value in update_data.items():
            if key in forbidden_keys or key.startswith("__"):
                continue
            product[key] = value
            
        return {"message": "Product updated", "product": product}

    @staticmethod
    async def delete_product(product_id: int) -> Dict[str, Any]:
        await asyncio.sleep(0.1)
        
        for i, p in enumerate(products):
            if p["id"] == product_id:
                deleted = products.pop(i)
                return {"message": "Product deleted", "deleted": deleted}
                
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Product not found")

    @staticmethod
    def export_products(min_price: Optional[float] = None, max_price: Optional[float] = None) -> Dict[str, Any]:
        try:
            results = list(products)
            if min_price is not None:
                results = [p for p in results if p["price"] >= min_price]
            if max_price is not None:
                results = [p for p in results if p["price"] <= max_price]
            return {"message": "Export success", "results": results}
        except Exception as error:
            print("Export failed:", str(error))
            raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail="Failed to export products")
